source/packages/services/assetlibrary/src/data/base.full.dao.ts

```diff
@@ -18,23 +18,40 @@ import { TYPES } from '../di/types';
 @injectable()
 export class BaseDaoFull {
     private _graph: structure.Graph;
+    private _conn: driver.DriverRemoteConnection | null;
 
     public constructor(
         @inject('neptuneUrl') private neptuneUrl: string,
         @inject(TYPES.GraphSourceFactory) graphSourceFactory: () => structure.Graph
     ) {
         this._graph = graphSourceFactory();
+        this._conn = null;
     }
 
-    protected getConnection(): NeptuneConnection {
+    protected async getConnection(): Promise<NeptuneConnection> {
         logger.debug(`base.full.dao getConnection: in:`);
-        const conn = new driver.DriverRemoteConnection(this.neptuneUrl, {
-            mimeType: 'application/vnd.gremlin-v2.0+json',
-            pingEnabled: false,
-        });
+
+        if (this._conn == null) {
+            logger.debug(`base.full.dao getConnection: create new connection:`);
+            this._conn = new driver.DriverRemoteConnection(this.neptuneUrl, {
+                mimeType: 'application/vnd.gremlin-v2.0+json',
+                pingEnabled: false,
+                connectOnStartup: false,
+            });
+            this._conn.addListener('close', (code: number, message: string) => {
+                logger.info(`base.full.dao connection close: code: ${code}, message: ${message}`);
+                this._conn = null;
+                if (code === 1006) {
```

Comment:
Other close codes indicate results like the user intentionally closing the connection, but 1006 is a code that occurs due to an unintended disconnection. Even when this happens, it's not handled by gremlin, and it ends up waiting until the Lambda times out. Therefore, an error has to be thrown in this case.

Reply:
@aaronatbissell sorry that this is happening. Can you share more info about what you're seeing? Error 1006 is an underlying websocket issue that cannot be recovered from. As @yuma124 mentions, an error is thrown to fail this faster, but ideally, this should not be a common issue.
- Does it fail 100% of the time or is it inconsistent?
- Can you add logging to discover which calls cause the error?
- It may be possible to get more info on where it's stuck by commenting out the `throw` so it hangs.
- Are you seeing any errors from Neptune?
- On the rest API side, you should catch this and retry it, which also might help identifying which call is triggering it.
Also, it might be helpful to get any information off the socket connection that might tell you the state of it (e.g. you should be able to get the current number of connections https://stackoverflow.com/questions/10275667/socket-io-connected-user-count ).

Reply:
> @aaronatbissell sorry that this is happening. Can you share more info about what you're seeing? Error 1006 is an underlying websocket issue that cannot be recovered from. As @yuma124 mentions, an error is thrown to fail this faster, but ideally, this should not be a common issue.
> 
>     * Does it fail 100% of the time or is it inconsistent?
It's very inconsistent as to when it happens, but it happens often
> 
>     * Can you add logging to discover which calls cause the error?
I've seen this in a couple of different calls. `getGroup` and `createGroup` stick out as more often, but that's probably just because those are the calls we perform the most often
> 
>     * It may be possible to get more info on where it's stuck by commenting out the `throw` so it hangs.
> 
>     * Are you seeing any errors from Neptune?
I'm not sure exactly what kind of errors you are looking for, but CloudWatch shows lots of "Client Errors per second" during that time frame:
![image](https://github.com/aws/aws-connected-device-framework/assets/89797921/ee03e5a3-5b80-4faa-81f0-f143ad2e4f45)

> 
>     * On the rest API side, you should catch this and retry it, which also might help identifying which call is triggering it.
>       Also, it might be helpful to get any information off the socket connection that might tell you the state of it (e.g. you should be able to get the current number of connections https://stackoverflow.com/questions/10275667/socket-io-connected-user-count ).

Yesterday we rolled back to v2.4.0 (before this PR was merged in) at about 17:00 (local time). I threw together a cloudwatch graph that shows a couple of interesting things:
1. The average number of open Gremlin websocket connections (blue) is actually MUCH higher with version 2.5.2 (before 17:00)
2. The number of assetlibrary-rest lambda failures (red) is much higher with version 2.5.2
3. The spikes you see after 17:00 is our normal test suite running, which stresses the system for about 30 minutes at a time. We were running this same test suite before 17:00 with many errors. We then had 0 errors after 17:00

![image](https://github.com/aws/aws-connected-device-framework/assets/89797921/aeb521f4-5512-4119-872c-8d3ca5baa22f)

Reply:
A couple of other baselines with out setup:

1. We are on Neptune Engine Version 1.1.1.0
2. We are using a db.r5.large instance size
3. We have a read replica set up
4. We are using a pretty vanilla version of CDF (very limited changes)

Reply:
Ok - I just saw #178 - I'm wondering if this is the same errors?

@yuma124 @ts-amz

Reply:
Yes, it appears to be the same error as described in #178.

Reply:
@aaronatbissell have you used the fix in https://github.com/aws/aws-connected-device-framework/pull/178 and does it solve this connection issue?

Reply:
I'm just getting back to this. We haven't tried to fix this problem but would really like to see this fixed so we can upgrade asset library.

Reply:
@aaronatbissell have you tried the update in https://github.com/aws/aws-connected-device-framework/pull/178 and validated whether or not it fixes the issues you are seeing? If not, are you able to test it?

Reply:
I haven't tested #178 but, as I [commented](https://github.com/aws/aws-connected-device-framework/pull/178#issuecomment-1912109466) in the PR, I think a better solution would be to open the neptune connection outside of the lambda handler. 

As far as testing goes, I don't plan on getting around to test this any time soon....sorry!